test/TemplateTests.targets.in

```diff
@@ -109,34 +109,46 @@ Copyright (c) .NET Foundation. All rights reserved.
         <_ExplicitAspNetCoreAllReference Include="@(_AspNetCoreAllReference->HasMetadata('Version'))" />
         <_AspNetCoreAppReference Include="@(PackageReference->WithMetadataValue('Identity', '$(_AspNetCoreAppPackageName)'))" />
         <_ExplicitAspNetCoreAppReference Include="@(_AspNetCoreAppReference->HasMetadata('Version'))" />
+      </ItemGroup>
 
-        <!-- Update implicit ASPNET metapackage references if needed -->
-        <PackageReference
-          Update="$(_AspNetCoreAllPackageName)"
-          Condition="'@(_AspNetCoreAllReference->Count())' == '1' AND '@(_ExplicitAspNetCoreAllReference->Count())' == '0'">
+      <!-- Update Microsoft.AspNetCore.All references -->
+      <ItemGroup Condition="'@(_AspNetCoreAllReference->Count())' == '1' AND '@(_ExplicitAspNetCoreAllReference->Count())' == '0'">
+        <!-- Update implicit ASPNET metapackage reference versions -->
+        <PackageReference Update="$(_AspNetCoreAllPackageName)">
           <Version>$(AspNetCoreAllRuntimeFrameworkVersion)</Version>
           <IsImplicitlyDefined>true</IsImplicitlyDefined>
+        </PackageReference>
+
+        <!-- Mark implicit ASPNET metapackage reference with PrivateAssets if not executable -->
+        <PackageReference
+          Update="$(_AspNetCoreAllPackageName)"
+          Condition="'$(OutputType)' != 'Exe' AND '$(_TargetFrameworkVersionWithoutV)' != '' AND '$(_TargetFrameworkVersionWithoutV)' >= '2.0'">
           <PrivateAssets>All</PrivateAssets>
           <Publish>true</Publish>
         </PackageReference>
-        <PackageReference
-          Update="$(_AspNetCoreAppPackageName)"
-          Condition="'@(_AspNetCoreAppReference->Count())' == '1' AND '@(_ExplicitAspNetCoreAppReference->Count())' == '0'">
+
+        <!-- Update list of implicit platform packages which require version verification -->
+        <ExpectedPlatformPackages Include="$(_AspNetCoreAllPackageName)" ExpectedVersion="$(AspNetCoreAllRuntimeFrameworkVersion)"/>
+      </ItemGroup>
+
+      <!-- Update Microsoft.AspNetCore.All references -->
+      <ItemGroup Condition="'@(_AspNetCoreAppReference->Count())' == '1' AND '@(_ExplicitAspNetCoreAppReference->Count())' == '0'">
+        <!-- Update implicit ASPNET metapackage reference versions -->
+        <PackageReference Update="$(_AspNetCoreAppPackageName)">
           <Version>$(AspNetCoreAppRuntimeFrameworkVersion)</Version>
           <IsImplicitlyDefined>true</IsImplicitlyDefined>
+        </PackageReference>
+
+        <!-- Mark implicit ASPNET metapackage reference with PrivateAssets if not executable -->
+        <PackageReference
+          Update="$(_AspNetCoreAppPackageName)"
+          Condition="'$(OutputType)' != 'Exe' AND '$(_TargetFrameworkVersionWithoutV)' != '' AND '$(_TargetFrameworkVersionWithoutV)' >= '2.0'">
```

Comment:
Conditions are taken from https://github.com/dotnet/sdk/blob/release/2.1.3xx/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.props#L54.

Reply:
I don't think the same conditions apply. The conditions for setting Microsoft.NETCore.App as PrivateAssets=All are:

 1. TFM >= netcoreapp2.0
 1. OutputType = library

We don't have to worry about the first condition - we already have the TFM guardrails in the package itself. I think we could just drop that.

The second condition deserves more conversation. I don't think we should do this. The reason the SDK can mark the package as PrivateAssets all is that they have their own TFM. "netcoreapp2.0" is what should define the API set and the Microsoft.NETCore.App package is mostly be an implementation detail of how that API is made available to the compiler.

ASP.NET Core, on the other hand, does not have a TFM by which we can define our API. Instead, NuGet packages are the way we define available API sets. If we make the PackageReference PrivateAssets=All, anyone writing a netcoreapp2.1 class library will end up with under-defined dependencies. Projects referencing this will end up with missing types at runtime. This includes projects referencing an aspnetcore class library as  ProjectReference, and external consumers using it as a PackageReference.

cc @davidfowl @DamianEdwards

Reply:
Curious why we wanted to add the PrivateAsset=All in the first place then? https://github.com/aspnet/websdk/pull/330#discussion_r182483906.

Reply:
I can't remember why I thought that the reference should have `PrivateAssets="all"`, but based on this (and Nate's reasoning here), I think I was wrong.

Reply:
We talked about this more in person. We actually want to keep PrivateAssets="All" on Microsoft.AspNetCore.App and .All under all conditions, not just OutputType==library. The primary reason for this thinking about how transitive references through p2p work. 

Consider the scenario where PrivateAssets != "all".
```
WebApp1 
  -PkgRef-> Microsoft.AspNetCore.App

TestProj1
  -ProjRef-> ..\WebApp1\WebApp1.csproj
```

In this scenario, NuGet will import Microsoft.AspNetCore.App's build assets into WebApp1, but not TestProj1. It will import its references into both. This breaks running on the ASP.NET Core shared runtime because the build assets from Microsoft.AspNetCore.App are required to set MicrosoftNETPlatformLibrary.

Now, when PrivateAssets=all is set, TestProj1 gets neither references nor build assets from Microsoft.AspNetCore.App. Although it's still less than ideal, we think this is preferable. We hope it will make it clearer that users need an explicit Microsoft.AspNetCore.App reference in all projects that expect to run on the ASP.NET Core shared runtime.